output/cloud/expv2/collect.go

```diff
@@ -0,0 +1,168 @@
+package expv2
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"go.k6.io/k6/metrics"
+)
+
+type timeBucket struct {
+	// TODO: for performance reasons, use directly a unix time here
+	// so we can avoid time->unix->time
+	Time  time.Time
+	Sinks map[metrics.TimeSeries]metrics.Sink
+}
+
+// bucketQ is a queue for buffering the aggregated metrics
+// that have to be flushed. It is expected to be used concurrently.
+type bucketQ struct {
+	m       sync.Mutex
+	buckets []timeBucket
+}
+
+// PopAll returns a slice with all the pushed buckets.
+// It returns a nil slice if the queue is empty.
+func (q *bucketQ) PopAll() []timeBucket {
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	if len(q.buckets) < 1 {
+		return nil
+	}
+
+	// return the enqueued slice and its relative array and allocate a new one
+	// using the same capacity.
+	b := q.buckets
+	q.buckets = make([]timeBucket, 0, len(b))
+	return b
+}
+
+// Push enqueues a new item in the queue.
+func (q *bucketQ) Push(b []timeBucket) {
+	if len(b) < 1 {
+		return
+	}
+	q.m.Lock()
+	q.buckets = append(q.buckets, b...)
+	q.m.Unlock()
+}
+
+type collector struct {
+	bq      bucketQ
+	nowFunc func() time.Time
+
+	aggregationPeriod time.Duration
+	waitPeriod        time.Duration
+
+	// We should no longer have to handle metrics that have times long in the past.
+	// So, instead of a map, we can probably use a simple slice (or even an array!)
+	// as a ring buffer to store the aggregation buckets.
+	// This should save us a some time, since it would make the lookups and
+	// WaitPeriod checks basically O(1).
+	// And even if for some reason there are occasional metrics with past times
+	// that don't fit in the chosen ring buffer size, we could just send them
+	// along to the buffer unaggregated.
+	timeBuckets map[int64]map[metrics.TimeSeries]metrics.Sink
+}
+
+func newCollector(aggrPeriod, waitPeriod time.Duration) (*collector, error) {
+	if aggrPeriod == 0 {
+		return nil, errors.New("aggregation period is not allowed to be zero")
+	}
+	if waitPeriod == 0 {
+		// TODO: we could simplify the expiring logic
+		// just having an internal static logic.
+		// Like skip only not closed buckets bucketEnd > now.
+		return nil, errors.New("aggregation wait period is not allowed to be zero")
+	}
+	return &collector{
+		bq:                bucketQ{},
+		nowFunc:           time.Now,
+		timeBuckets:       make(map[int64]map[metrics.TimeSeries]metrics.Sink),
+		aggregationPeriod: aggrPeriod,
+		waitPeriod:        waitPeriod,
+	}, nil
+}
+
+// CollectSamples drain the buffer and collect all the samples.
+func (c *collector) CollectSamples(containers []metrics.SampleContainer) {
```

Comment:
Minor non-blocking nit, but is there a reason for some of the methods of `collector` and `bucketQ` to be exported? They are only used in the `expv2.Output`, right? So it seems like everything could be private for just this package?

Reply:
I set a `TODO:` for this to check all the export/unexport states and add it to one of the last PRs.